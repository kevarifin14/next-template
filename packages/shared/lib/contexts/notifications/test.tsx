import { act, fireEvent, render } from '@testing-library/react';
import React from 'react';

import { NotificationsProvider, useNotifications } from '.';

const mockId = '1234';
const mockTitle = 'mockTitle';
const mockDescription = 'mockDescription';

const mockNotification = {
  id: mockId,
  title: mockTitle,
  description: mockDescription,
};

const mockAddNotificationButtonText = 'mockAddNotificationButtonText';
const mockAddSuccessNotificationButtonText = 'mockAddSuccessNotificationButtonText';
const mockAddErrorNotificationButtonText = 'mockAddErrorNotificationButtonText';

const MockNotificationConsumer = () => {
  const {
    addNotification,
    addErrorNotification,
    addSuccessNotification,
  } = useNotifications();

  return (
    <>
      <button onClick={() => addNotification(mockNotification)}>
        {mockAddNotificationButtonText}
      </button>

      <button onClick={() => addSuccessNotification(mockNotification)}>
        {mockAddSuccessNotificationButtonText}
      </button>

      <button onClick={() => addErrorNotification(mockNotification)}>
        {mockAddErrorNotificationButtonText}
      </button>
    </>
  );
};

const MockNotificationProvider = () => (
  <NotificationsProvider>
    <MockNotificationConsumer />
  </NotificationsProvider>
);

describe('NotificationProvider', () => {
  it('matches snapshot', () => {
    const { baseElement } = render(<MockNotificationProvider />);
    expect(baseElement).toMatchSnapshot();
  });

  it('adds notification and removes after 4 seconds', () => {
    jest.useFakeTimers();

    const { queryByText } = render(<MockNotificationProvider />);

    const mockAddNotificationButton = queryByText(mockAddNotificationButtonText);

    fireEvent.click(mockAddNotificationButton);

    act(() => {
      jest.advanceTimersByTime(4100);
    });

    expect(queryByText(mockTitle)).toBeNull();
  });

  it('adds success notification', () => {
    const { queryByText, getByRole } = render(<MockNotificationProvider />);

    const mockAddSuccessNotificationButton = queryByText(mockAddSuccessNotificationButtonText);

    fireEvent.click(mockAddSuccessNotificationButton);

    expect(queryByText(mockTitle)).toBeTruthy();

    const mockNotificationCloseButton = getByRole('button', { name: 'Close' });

    fireEvent.click(mockNotificationCloseButton);

    expect(queryByText(mockTitle)).toBeNull();
  });

  it('adds error notification', () => {
    const { queryByText } = render(<MockNotificationProvider />);

    const mockAddErrorNotificationButton = queryByText(mockAddErrorNotificationButtonText);

    fireEvent.click(mockAddErrorNotificationButton);

    expect(queryByText(mockTitle)).toBeTruthy();
  });
});
