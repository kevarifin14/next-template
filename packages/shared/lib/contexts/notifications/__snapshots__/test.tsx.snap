// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationProvider matches snapshot 1`] = `
<body>
  <div>
    <div
      class="fixed max-w-sm w-full right-0 top-0 flex flex-col justify-end items-start space-y-4 pointer-events-none z-30 p-4"
    />
    <button>
      mockAddNotificationButtonText
    </button>
    <button>
      mockAddSuccessNotificationButtonText
    </button>
    <button>
      mockAddErrorNotificationButtonText
    </button>
  </div>
</body>
`;
